src/common/main/kotlin/tech/thatgravyboat/skyblockapi/api/data/StoredFolderData.kt

```diff
@@ -0,0 +1,73 @@
+package tech.thatgravyboat.skyblockapi.api.data
+
+
+import com.mojang.serialization.Codec
+import tech.thatgravyboat.skyblockapi.api.SkyBlockAPI
+import tech.thatgravyboat.skyblockapi.helpers.McClient
+import java.nio.file.Path
+import kotlin.io.path.*
+
+internal class FolderStorage<T : Any>(
+    val folder: String,
+    val default: T,
+    val codec: Codec<T>,
+) {
+    private val storages = mutableMapOf<String, StoredData<T>>()
+    private val defaultPath: Path = McClient.config.resolve("skyblockapi/$folder")
+
+    init {
+        load()
+    }
+
+    fun load() {
+        this.storages.putAll(
+            files().mapNotNull {
+                val id = it.nameWithoutExtension
+                try {
+                    id to StoredData(
+                        version = 0,
+                        data = default,
+                        file = "$folder/$id.json",
+                        codec = { codec },
+                    )
+                } catch (e: Exception) {
+                    SkyBlockAPI.error("Failed to load storage file: ${it.relativeTo(McClient.config)}", e)
+                    null
+                }
+            },
+        )
+    }
+
+    fun add(value: T) = set(value.hashCode().toString(), value)
+
+    fun set(id: String, value: T) {
+        storages.getOrPut(id) {
+            StoredData(
+                version = 0,
+                dataProvider = { value },
+                file = "$folder/$id.json",
+                codec = { codec },
+            )
+        }.apply {
+            set(value)
+        }
```

Comment:
```suggestion
        }.set(value)
```

or can be put into a variable and then .set but I dont like the idea of using apply here